Replace databaseInterface with Django models

mistydemeo opened the pull request:
~~`databaseInterface` is still imported in a few places in MCPServer, primarily for the `getUTCTime` function that will be moved elsewhere, and for database locking that may be removeable.~~

See comments below.

Commit:
MCPServer: replace databaseInterface with django models

Commit:
Unit classes: move some methods into superclass

Also switch from databaseInterface to Django models.

refs #7137.

mistydemeo commented on src/MCPServer/lib/archivematicaMCP.py:
```diff
@@ -114,25 +114,18 @@ def findOrCreateSipInDB(path, waitSleep=dbWaitSleep):
-        while row != None:
-            UUID = row[0]
+        sips = SIP.objects.filter(currentpath=path)
+        if len(sips) > 1:
```
In most places I've tried to not acknowledge the possibility of multiple results (unlike the original code) - but here there is a vague chance that this could happen.

Hwesta commented on src/MCPServer/lib/archivematicaMCP.py:
```diff
@@ -114,25 +114,18 @@ def findOrCreateSipInDB(path, waitSleep=dbWaitSleep):
-        sql = """SELECT sipUUID FROM SIPs WHERE sipUUID = '""" + UUID + "';"
-        rows = databaseInterface.queryAllSQL(sql)
-        if not rows:
+        if not SIP.objects.filter(uuid=UUID):
```
If we're not evaluating the queryset, [`.exists()`](https://docs.djangoproject.com/en/dev/ref/models/querysets/#exists) seems to be good to use in the query.

mistydemeo replied:
Fixed!

Hwesta commented on src/MCPServer/lib/archivematicaMCP.py:
```diff
@@ -114,25 +114,18 @@ def findOrCreateSipInDB(path, waitSleep=dbWaitSleep):
+            print "Warning: found more than one matching SIP for path", path, "- using first result"
+        if sips:
+            UUID = sips[0].uuid
             print "Opening existing SIP:", UUID, "-", path
```
I prefer not to use all caps for anything but constants - maybe sip_uuid for this variable?

mistydemeo replied:
Good point - seems like a decent chance to improve some of the variable naming.

Hwesta commented on src/MCPServer/lib/archivematicaMCP.py:
```diff
@@ -196,17 +189,14 @@ def createUnitAndJobChainThreaded(path, config, terminate=True):
+    watched_directories = WatchedDirectory.objects.all()
+
+    for dir in watched_directories:
+        directory = dir.watched_directory_path.replace("%watchDirectoryPath%", config.get('MCPServer', "watchDirectoryPath"), 1)
```
Move querying the config for watchDirectoryPath outside the loop?  I don't know how efficient config.get is.

mistydemeo replied:
Moved outside the loop.

Hwesta commented on src/MCPServer/lib/jobChainLink.py:
```diff
@@ -37,19 +37,21 @@
+constOneTask = TaskType.objects.get(description="one_instance").uuid
+constTaskForEachFile = TaskType.objects.get(description="for each file").uuid
+constSelectPathTask = TaskType.objects.get(description="get user choice to proceed with").uuid
+constSelectPathTask = TaskType.objects.get(description="get user choice to proceed with").uuid
```
Duplicate

Hwesta commented on src/MCPServer/lib/jobChainLink.py:
```diff
@@ -62,27 +64,17 @@ def __init__(self, jobChain, jobChainLinkPK, unit, passVar=None, subJobOf=""):
-        sql = """SELECT MicroServiceChainLinks.currentTask, MicroServiceChainLinks.defaultNextChainLink, TasksConfigs.taskType, TasksConfigs.taskTypePKReference, TasksConfigs.description, MicroServiceChainLinks.reloadFileList, Sounds.fileLocation, MicroServiceChainLinks.defaultExitMessage, MicroServiceChainLinks.microserviceGroup FROM MicroServiceChainLinks LEFT OUTER JOIN Sounds ON MicroServiceC ...
-        c, sqlLock = databaseInterface.querySQL(sql)
-        row = c.fetchone()
-        if row == None:
```
Do we need the None handling?  Try/except that returns None around the get?

mistydemeo replied:
This is in the deleted part of the diff!

mistydemeo replied:
Dang, turns out this really can return no value. :| Adding a check.

Hwesta commented on src/MCPServer/lib/jobChainLink.py:
```diff
@@ -120,47 +112,35 @@ def createTasks(self, taskType, taskTypePKReference):
         else:
             print sys.stderr, "unsupported task type: ", taskType
 
+    # Deprecated, remove later
```
:+1:

Hwesta commented on src/MCPServer/lib/jobChainLink.py:
```diff
@@ -120,47 +112,35 @@ def createTasks(self, taskType, taskTypePKReference):
+            try:
+                link = MicroServiceChainLinkExitCode.objects.get(microservicechainlink=str(self.pk), exitcode=str(exitCode))
+                return link.pk
+            except MicroServiceChainLinkExitCode.DoesNotExist:
```
Should catch `MicroServiceChainLinkExitCode.MultipleObjectsReturned` as well in case someone messed up the MSCLEC table.

mistydemeo replied:
Updated.

Hwesta commented on src/MCPServer/lib/jobChainLink.py:
```diff
@@ -120,47 +112,35 @@ def createTasks(self, taskType, taskTypePKReference):
+                link = MicroServiceChainLinkExitCode.objects.get(microservicechainlink=str(self.pk), exitcode=str(exitCode))
+                return link.pk
+            except MicroServiceChainLinkExitCode.DoesNotExist:
+                return self.defaultNextChainLink
```
Add an explicit return at the bottom? As is I don't if None is correct if the 'if' block isn't hit (though I think None is correct in this case)

mistydemeo replied:
It's the original behaviour, so I think it's fine for it to not return in this case?

Hwesta replied:
-shrug- Fine to leave it, but I prefer an explicit return, if elsewhere it does explicit return.

Hwesta commented on src/MCPServer/lib/jobChainLink.py:
```diff
@@ -120,47 +112,35 @@ def createTasks(self, taskType, taskTypePKReference):
         else:
             print "No exit message"
 
+    # Deprecated, remove later
```
This should be moved to the playSounds block, since other things in this function are important.

Hwesta commented on src/MCPServer/lib/linkTaskManagerChoice.py:
```diff
@@ -48,16 +49,10 @@ def __init__(self, jobChainLink, pk, unit):
-            print row
-            chainAvailable = row[0]
-            description = row[1]
-            self.choices.append((chainAvailable, description,))
```
This got dropped in changes - is it important?

mistydemeo replied:
Whoops, looks like I omitted the loop. Updated.

Hwesta commented on src/MCPServer/lib/linkTaskManagerDirectories.py:
```diff
@@ -28,41 +28,27 @@
-        sql = """SELECT * FROM StandardTasksConfigs where pk = '%s'""" % (pk.__str__())
-        c, sqlLock = databaseInterface.querySQL(sql)
-        row = c.fetchone()
-        if row == None:
```
Handle this in rewrite?  Or is the raised Exception from get enough?

mistydemeo replied:
I think the exception is probably fine, and more explicit. I don't know that this original condition ever happened.

Hwesta replied:
It shouldn't!  And the Exception is probably clearer anyway.

Hwesta commented on src/MCPServer/lib/linkTaskManagerFiles.py:
```diff
@@ -45,40 +45,39 @@ def __init__(self, jobChainLink, pk, unit):
+            var = UnitVariable.objects.get(unittype=self.unit.unitType,
+                                           unituuid=self.unit.UUID,
+                                           variable=self.execute)
+        except UnitVariable.DoesNotExist:
```
Catch MultipleObjectsReturned as well?

Hwesta commented on src/MCPServer/lib/unit.py:
```diff
@@ -27,6 +27,38 @@ def __init__(self, currentPath, UUID):
+                                               archivematicaMCP.config.get('MCPServer', "sharedDirectory"), 1) + "/"
+        try:
+            for directory, subDirectories, files in os.walk(currentPath):
+                directory = directory.replace( currentPath, self.pathString, 1) 
```
unitDIP needs pathString set

mistydemeo replied:
Updated.

Hwesta commented on src/MCPServer/lib/unit.py:
```diff
@@ -34,7 +66,29 @@ def setMagicLink(self,link, exitStatus=""):
+            )
+            var.save()
     
     def getmicroServiceChainLink(self, variable, variableValue, defaultMicroServiceChainLink):
```
`variableValue` is not used here (or in the original) - can it be removed? Or a comment reminding that work to happen?

mistydemeo replied:
I don't want to change the method signature (even though this is 100% internal), but I'll add a comment.

Hwesta commented on src/MCPServer/lib/unitTransfer.py:
```diff
@@ -49,143 +51,54 @@ def __init__(self, currentPath, UUID=""):
-            row = c.fetchone()
-            while row != None:
-                UUID = row[0]
-                print "Opening existing Transfer:", UUID, "-", currentPath2
```
Keep this print statement in the rewrite?

Hwesta commented on src/dashboard/src/main/models.py:
```diff
@@ -586,7 +595,13 @@ class Meta:
+    filterfileend = models.CharField(max_length=50, null=True, blank=True, db_column='filterFileEnd')
+    filterfilestart = models.CharField(max_length=50, null=True, blank=True, db_column='filterFileStart')
+    filtersubdir = models.CharField(max_length=50, null=True, blank=True, db_column='filterSubDir')
+    requiresoutputlock = models.IntegerField(null=True, default=None, db_column='requiresOutputLock')
```
I think this is actually a BooleanField (or NullBooleanField).  Django's BooleanField [maps to tinyint](http://stackoverflow.com/questions/1778948/boolean-fields-in-mysql-django-models).

mistydemeo replied:
I noticed another tinyint boolean was mapped to an IntegerField, so I wasn't sure if we had a preference. I can move these over to BooleanFields though.

Hwesta replied:
Since it's semantically a boolean, I think it makes sense to have the field reflect that.  I think BooleanField is more correct for this case, but since the DB allows NULL, then NullBooleanField may be better?  OTOH, if we want to eventually generate the tables from the models, BooleanField is what we're really want.

Hwesta commented on src/dashboard/src/main/models.py:
```diff
@@ -605,6 +620,49 @@ class Meta:
+
+class TaskConfigSetUnitVariable(models.Model):
+    id = UUIDPkField()
+    variable = models.TextField(null=True, blank=True)
```
Include db_column.  Django will guess the correct value here, but the explicit is nice, especially since everything else has it.  Same on TaskConfigUnitVariableLinkPull.

Hwesta commented on src/dashboard/src/main/models.py:
```diff
@@ -688,3 +746,24 @@ class Meta:
+    id = UUIDPkField()
+    watched_directory_path = models.TextField(null=True, blank=True, db_column='watchedDirectoryPath')
+    chain = models.ForeignKey('MicroServiceChain', null=True)
+    only_act_on_directories = models.IntegerField(default=1, db_column='onlyActOnDirectories')
```
Also a [Null]BooleanField.

Hwesta commented on src/MCPServer/lib/jobChainLink.py:
```diff
@@ -37,19 +37,21 @@
+constOneTask = TaskType.objects.get(description="one instance").pk
+constTaskForEachFile = TaskType.objects.get(description="for each file").pk
+constSelectPathTask = TaskType.objects.get(description="get user choice to proceed with").pk
+constSelectPathTask = TaskType.objects.get(description="get user choice to proceed with").pk
```
Dupe of above line.

mistydemeo commented:
(Will squash down commits after review.)

Hwesta committed:
MCPClient: Remove DashboardSettings dependency

Giving the MCPClient access to the dashboard UUID via replacement dict
& environment variable was a branch of code development that was not
followed up on. Remove it to reduce table dependencies of the MCPClient.

Hwesta committed:
Models: Update to match SQL

refs #7303

Commit:
databaseFunctions: use Django models

refs #7137

Commit:
fileOperations: use Django models

refs #7137

Commit:
elasticSearchFunctions: use Django models

refs #7137

Commit:
storageService: use Django models

refs #7137

Commit:
MCPClient: use Django models

refs #7137

mistydemeo commented:
Updated to replace databaseInterface in archivematicaCommon too, and rebased on top of dev/issue-7303-update-main-models to incorporate all of those changes.

Commit:
jobChainLink: print to stderr, don't print stderr

Commit:
localDevSetup: replace databaseInterface with Django models

Commit:
identifyFileFormat: file_.sip won't raise

Commit:
databaseFunctions: add docstrings

Commit:
createSIPfromTransferObjects: fix subsequent runs

If this script runs more than once, the SIP objects directory will
already be populated with some files. Make sure the contents of
objects/ is a directory before trying to list its contents, otherwise
this will fail.

Commit:
databaseFunctions: add tests

Commit:
dicts: don't try to expand path if sipdir not defined

Commit:
insertIntoDerivations: source and derived files are mandatory

Commit:
main.models: add FilesIDs model

Commit:
Move getUTCDate, getDeciDate to databaseFunctions

refs #7137

Commit:
databaseInterface: remove separator

Commit:
views_atk: remove unused import

refs #4757

mistydemeo commented:
Squashed and rebased on qa/1.x. This is ready for review.

At this point databaseInterface has been replaced throughout MCPServer and MCPClient, along with localDevSetup; it remains in use in the following places:
- databaseInterface sqlLock in MCPServer and RCPServer - this can probably be removed?
- Elasticsearch backup
- Unused code (archivematicaFITS, archivematicaFIDO, Tika)
- Archivist's Toolkit integration
- Normalization report (uses django.db.connection, not databaseInterface, but uses raw SQL for a complex statement)
- Email fail report (uses MySQL views that aren't in Django models)
- Maildir
- PlantUMLgenerator

mistydemeo commented:
Rebased on top of qa/1.x.

Hwesta commented on src/MCPClient/lib/archivematicaClient.py:
```diff
@@ -111,17 +96,17 @@ def executeCommand(gearman_worker, gearman_job):
-        rows = databaseInterface.queryAllSQL(sql)
-        if len(rows):
+        task = Task.objects.get(taskuuid=gearman_job.unique)
+        if task.starttime is not None:
```
Should this be catching a `DoesNotExist` error instead of checking starttime??  It looks like the old code checked for existence.

mistydemeo replied:
I think the old code checked for existence plus start time being non-zero; the logging below indicates it primarily cares about the latter (since the task should always exist here). The implicit assumption appeared to be that there would always be a Task with the requested taskUUID.

Hwesta replied:
Ok

Hwesta commented on src/MCPClient/lib/clientScripts/archivematicaAssignFileUUID.py:
```diff
@@ -53,8 +55,9 @@
-            databaseInterface.runSQL(sql)
+            f = File.objects.get(uuid=fileUUID)
+            f.filegrpuse = opts.use
+            f.save()
```
Use [update](https://docs.djangoproject.com/en/1.5/ref/models/querysets/#update) for this?

mistydemeo replied:
Updated.

There's... probably quite a few places I could fix here. 😭

Hwesta commented on src/MCPClient/lib/clientScripts/archivematicaCreateMETS.py:
```diff
@@ -119,15 +121,18 @@ def createFileSec(path, parentBranch, structMapParent):
+                        myuuid = f.uuid
 
                     if includeAmdSec:
                         createDigiprovMD(myuuid, itempath, myuuid)
```
It's a good thing `includeAmdSec` must never be true, since `createDigiprovMD` isn't and wasn't imported.

mistydemeo replied:
😱

Hwesta commented on src/MCPClient/lib/clientScripts/archivematicaCreateMETS.py:
```diff
@@ -119,15 +121,18 @@ def createFileSec(path, parentBranch, structMapParent):
+                        f = File.objects.get(**kwargs)
+                    except File.DoesNotExist:
+                        print >> sys.stderr, "No uuid for file: \"", pathSTR, "\""
+                        continue
```
Should this `continue` be here?  Should it fail the microservice if it cannot find an associated file or continue with a default `myuuid` value or skip this file?

mistydemeo replied:
I guess for consistency with the old behaviour it shouldn't continue, just proceed with the empty string.

mistydemeo replied:
Removed that line.

Hwesta commented on src/MCPClient/lib/clientScripts/archivematicaCreateMETS2.py:
```diff
@@ -130,25 +135,20 @@ def createAgent(agentIdentifierType, agentIdentifierValue, agentName, agentType)
 FileMetadataAppliesToType = '7f04d9d4-92c2-44a5-93dc-b7bfdf0c1f17'
 def getDublinCore(unit, id):
     field_list = ["title", "creator", "subject", "description", "publisher", "contributor", "date", "type", "format", "identifier", "source", "relation", "language", "coverage", "rights", "isPartOf"]
```
isPartOf should be fixed to is_part_of to reflect DB-SQL vs DB-Django.

Hwesta commented on src/MCPClient/lib/clientScripts/archivematicaCreateMETS2.py:
```diff
@@ -130,25 +135,20 @@ def createAgent(agentIdentifierType, agentIdentifierValue, agentName, agentType)
+        return
+
+    ret = etree.Element(ns.dctermsBNS + "dublincore", nsmap={"dc":ns.dctermsNS})
+    ret.set(ns.xsiBNS+"schemaLocation", ns.dctermsNS + " http://dublincore.org/schemas/xmls/qdc/2008/02/11/dcterms.xsd")
```
Fix linter errors on these lines?

mistydemeo replied:
Updated.

Hwesta commented on src/MCPClient/lib/clientScripts/archivematicaCreateMETS2.py:
```diff
@@ -290,14 +290,9 @@ def createTechMD(fileUUID):
-        row = c.fetchone()
-    sqlLock.release()
+    f = File.objects.get(uuid=fileUUID)
+    fileSize = f.size
```
This needs to be converted to a string.

mistydemeo commented:
Made some style improvements in MCPClient.

Hwesta commented on src/MCPClient/lib/clientScripts/archivematicaCreateMETSRights.py:
```diff
@@ -27,10 +27,12 @@
 import lxml.etree as etree
 
+sys.path.append("/usr/share/archivematica/dashboard")
+from main.models import RightsStatement, RightsStatementCopyright, RightsStatementCopyrightDocumentationIdentifier, RightsStatementCopyrightNote, RightsStatementLicense, RightsStatementLicenseDocumentationIdentifier, RightsStatementLicenseNote, RightsStatementLinkingAgentIdentifier, RightsStatementOtherRightsDocumentationIdentifier, RightsStatementOtherRightsInformation, RightsStatementOtherRight ...
```
Only `RightsStatement` is actually used - the rest don't need to be imported.

mistydemeo replied:
Good catch - they were used in a previous version and I didn't update the import.

Hwesta commented on src/MCPClient/lib/clientScripts/archivematicaCreateMETSRights.py:
```diff
@@ -46,180 +48,147 @@ def archivematicaGetRights(metadataAppliesToList, fileUUID):
+            metadataappliestotype_id=metadataAppliesToType
+        )
+        for statement in statements:
+            valueDic= {}
```
`valueDic` is never used.

Hwesta commented on src/MCPClient/lib/clientScripts/archivematicaCreateMETSRights.py:
```diff
@@ -46,180 +48,147 @@ def archivematicaGetRights(metadataAppliesToList, fileUUID):
+                        etree.SubElement(copyrightApplicableDates, ns.premisBNS + "endDate").text = formatDate(copyright.copyrightapplicableenddate)
+
+            elif statement.rightsbasis.lower() in ["license"]:
+                for license in rightsstatementlicense_set.all():
```
Should be `statement.rightsstatementlicense_set.all()`

Hwesta commented on src/MCPClient/lib/clientScripts/archivematicaCreateMETSRights.py:
```diff
@@ -46,180 +48,147 @@ def archivematicaGetRights(metadataAppliesToList, fileUUID):
+                    # otherRightsDocumentationIdentifier
+                    for identifier in info.rightsstatementotherrightsdocumentationidentifier_set.all():
+                        otherRightsDocumentationIdentifier = etree.SubElement(otherRightsInformation, ns.premisBNS + "otherRightsDocumentationIdentifier")
+                        etree.SubElement(otherRightsDocumentationIdentifier, ns.premisBNS + "otherRightsDocumentationIdentifierType").text = identiifer.otherrightsdocumentationidentifiertype
```
Typo `identiifer` -> `identifier`

Hwesta commented on src/MCPClient/lib/clientScripts/archivematicaCreateMETSRightsDspaceMDRef.py:
```diff
@@ -52,20 +54,19 @@ def createMDRefDMDSec(LABEL, itemdirectoryPath, directoryPathSTR):
+        mets = File.objects.get(currentlocation=path,
+                                transfer_id=transferUUID)
+        metsFileUUID = mets.uuid
+        metsloc = mets.currentlocation.replace("%SIPDirectory%", "", 1)
```
`metsloc` vs `metsLoc` variable name

Hwesta commented on src/MCPClient/lib/clientScripts/archivematicaAssignFileUUID.py:
```diff
@@ -53,8 +55,7 @@
         if opts.update_use:
-            sql = """UPDATE Files SET fileGrpUse='%s' WHERE fileUUID = '%s';""" % (opts.use, fileUUID)
-            databaseInterface.runSQL(sql)
+            File.objects.filter(uuid=fileUUID).update(filegrpuse=use)
```
`use` -> `opts.use`

Commit:
archivematicaMoveSIP: remove unused imports

Commit:
archivematicaMoveTransfer: remove unused imports

Commit:
unitSIP: remove unused imports

Commit:
unitTransfer: remove unused imports

Commit:
databaseFunctions: remove unused imports

Commit:
manualNormalizationRemoveMNDirectories: fix variable reference

Hwesta commented on src/MCPClient/lib/clientScripts/archivematicaCreateMETS2.py:
```diff
@@ -128,27 +133,24 @@ def createAgent(agentIdentifierType, agentIdentifierValue, agentName, agentType)
+
+
 def getDublinCore(unit, id):
     field_list = ["title", "creator", "subject", "description", "publisher", "contributor", "date", "type", "format", "identifier", "source", "relation", "language", "coverage", "rights", "isPartOf"]
```
`isPartOf` -> `is_part_of` to match Django model name not DB column name.

Hwesta commented on src/MCPClient/lib/clientScripts/archivematicaCreateMETSTrim.py:
```diff
@@ -83,47 +83,50 @@ def getTrimFileDmdSec(baseDirectoryPath, fileGroupIdentifier, fileUUID):
-        label = os.path.basename(row[0])
+    else:
+        label = os.path.basename(f.currentlocation)
         attrib = {"LABEL":label, ns.xlinkBNS + "href":row[0].replace("%SIPDirectory%", "", 1), "MDTYPE":"OTHER", "OTHERMDTYPE":"CUSTOM", 'LOCTYPE':"OTHER", 'OTHERLOCTYPE':"SYSTEM"}
```
`row[0]` should probably be something else.

Hwesta commented on src/MCPClient/lib/clientScripts/archivematicaCreateMETSTrim.py:
```diff
@@ -83,47 +83,50 @@ def getTrimFileDmdSec(baseDirectoryPath, fileGroupIdentifier, fileUUID):
+    try:
+        f = File.objects.get(removedtime__isnull=True,
+                             sip_id=fileGroupIdentifier,
+                             filegrpuuid=fileUUID)
```
This seems to have dropped the `fileGrpUse='TRIM file metadata'`

Hwesta commented on src/MCPClient/lib/clientScripts/backlogUpdatingTransferFileIndex.py:
```diff
@@ -21,19 +21,17 @@
-    databaseInterface.runSQL(sql)
+    Transfer.objects.filter(uuid=transferUUID).update(currentlocation=dst)
 
 def moveSIP(src, dst, transferUUID, sharedDirectoryPath):
```
moveSIP and updateDB aren't actually called. >__>  Can probably be removed.

mistydemeo replied:
Added a commit that removes them, and removed this script from the MCPClient commit.

Hwesta commented on src/MCPClient/lib/clientScripts/checkForAccessDirectory.py:
```diff
@@ -44,41 +47,30 @@ def something(SIPDirectory, accessDirectory, objectsDirectory, DIPDirectory, SIP
+                files = File.objects.filter(removedtime__isnull=True,
+                                            currentlocation__startswith=objectNameLike,
+                                            sip_id=SIPUUID)
+                if files.count() < 1:
```
`if not files.exists()` probably better

Commit:
backlogUpdatingTransferFileIndex: remove unused functions

mistydemeo commented:
Tested with many of the alternate workflows, with some additional bugfixes and stylistic improvements.

Commit:
createMETS2 Dspace: raise if no, or too many, originals found

Commit:
unit: currentPath must be a string for os.walk

mistydemeo commented:
At this point I've tested the following workflows to be working:
- Standard transfer
- Zipped bag
- Unzipped bag
- DSpace
- Manual normalization (via CSV, and via "normalize manually")
- CSV and JSON DC metadata
- Disk image with metadata
- Arbitrary non-Unicode filenames, and Unicode filenames

Hwesta commented on src/MCPClient/lib/clientScripts/createSIPfromTransferObjects.py:
```diff
@@ -76,14 +69,16 @@
-        fileUUID = row[0]
-        currentPath = databaseFunctions.deUnicode(row[1])
+    files = File.objects.filter(transfer_id=transferUUID,
+                                currentlocation__startswith='%transferDirectory%objects')
```
Should removedTime be in the query?

Hwesta commented on src/MCPClient/lib/clientScripts/createSIPfromTransferObjects.py:
```diff
@@ -76,14 +69,16 @@
+    files = File.objects.filter(transfer_id=transferUUID,
+                                currentlocation__startswith='%transferDirectory%objects')
+    for f in files:
+        fileUUID = f.uuid
```
fileUUID appears to be unused

Hwesta commented on src/MCPClient/lib/clientScripts/createSIPsfromTRIMTransferContainers.py:
```diff
@@ -63,23 +57,26 @@
+                                    currentlocation__startswith=dir,
+                                    transfer_id=transferUUID)
+        for f in files:
+            currentPath = databaseFunctions.deUnicode(f.currentlocation).replace('%transferDirectory%objects/' + container, '%transferDirectory%objects')
```
Replace `'%transferDirectory%objects/' + container` with variable defined above?  Rename `dir` so it doesn't conflict with builtins.

Hwesta commented on src/MCPClient/lib/clientScripts/emailFailReport.py:
```diff
@@ -25,24 +25,21 @@
-            continue
-        ret.append(email)
-    return ret
+    return [u[0] for u in User.objects.filter(is_active=True).values_list('email').exclude(email__in=['demo@example.com', ''])]
```
:+1:  I love it when we can replace 11 lines with 1 that's just as clear.

Hwesta commented on src/MCPClient/lib/clientScripts/loadLabelsFromCSV.py:
```diff
@@ -48,10 +48,4 @@
-
-    
-    
+            File.objects.filter(originallocation=label, transfer_id=transferUUID).update(label=label)
```
Should be `originallocation=filePath`

Hwesta commented on src/MCPClient/lib/clientScripts/manualNormalizationCheckForManualNormalizationDirectory.py:
```diff
@@ -40,7 +41,11 @@
+                                          unituuid=SIPUUID,
+                                          variable="returnFromManualNormalized")
+            uv.microservicechainlink_id = "f060d17f-2376-4c0b-a346-b486446e46ce"
+            uv.save()
```
Use `.update()`?

mistydemeo replied:
Updated - but this idiom is used all over the place (not just in this PR). Do you care enough about changing it all now?

Hwesta replied:
Kind of? It's better but depending on how many there are that might not justify the work involved.

mistydemeo replied:
Turns out there was a much smaller surface area of stuff to change here than I thought, and it's definitely better to be consistent at this point, so made the changes everywhere there wasn't much restructuring to do so.

Hwesta commented on src/MCPClient/lib/clientScripts/manualNormalizationCreateMetadataAndRestructure.py:
```diff
@@ -44,17 +46,21 @@
-    filePathLike2 = databaseInterface.MySQLdb.escape_string(filePathLike)[:-1]
-unitIdentifierType = "sipUUID"
+    filePathLike2 = filePathLike[:-1]
+unitIdentifierType = "sip_id"
```
Can we replace `unitIdentifierType` with just `sip_id`?  It only seems to have one value.

mistydemeo replied:
Good thought. I get the idea this might have been intended to be generalized like some of the other scripts, with this passed at the CLI, but since that didn't happen / won't happen I'd rather take it out.

mistydemeo replied:
Inlined both `sip_id` and used `SIPUUID` directly instead of `unitidentifier`. This also lets me avoid using kwargs dicts in favour of directly passing kwargs, so that's nice.

Hwesta commented on src/MCPClient/lib/clientScripts/manualNormalizationMoveAccessFilesToDIP.py:
```diff
@@ -39,21 +41,23 @@
-if len(rows) != 1:
+
+try:
+    path_condition = Q(currentlocation__startswith=filePathLike) | Q(currentlocation=filePathLike2)
```
While clever, I don't think that the Q statement does the same thing as the original code.

Given files `foo.txt` and `foobar.txt` with a search file of `foo.jpg`, the original code will only match `foo.txt` while the new code will match both and trigger the normalization.csv exception.  By searching for the path with the . first, and only searching without it if nothing was initially returned, we ensured that we'd find files that matched just the basename and reduced false positives.

Whether the old code was a _good_ way to accomplish our goal is another question...

mistydemeo replied:
Good catch - I copied the code here a bit too closely from the other tool that does something similar. Fixed.

mistydemeo commented on src/MCPClient/lib/clientScripts/archivematicaCreateMETS2.py:
```diff
@@ -638,28 +599,38 @@ def createFileSec(directoryPath, parentDiv):
-            sqlLock.release()
-            
+
+            kwargs = {
```
Oops, duplicated code in the merge.

Hwesta commented on src/MCPClient/lib/clientScripts/trimCreateRightsEntries.py:
```diff
@@ -106,74 +109,68 @@ def getDateTimeFromDateClosed(dateClosed):
+            rightsstatement=statement,
+            act="Disseminate",
+            startdate = startTime,
+            enddate = endTime
```
Nitpick, but remove spaces around the =?

Hwesta commented on src/MCPServer/lib/archivematicaMCP.py:
```diff
@@ -59,8 +58,8 @@
 from archivematicaFunctions import unicodeToStr
 
-# Print SQL sent to the database interface 
-databaseInterface.printSQL = True
```
Hmm, this is going to remove most of the logging in archivematicaMCPClient-<hostname>.log

... that's probably a good thing, but we should look at adding enough logging to usefully replace that.

Hwesta commented on src/MCPServer/lib/jobChain.py:
```diff
@@ -67,20 +64,11 @@ def __init__(self, unit, chainPK, notifyComplete=None, passVar=None, UUID=None,
-            row = c.fetchone()
-        sqlLock.release()
+
+        chain = MicroServiceChain.objects.get(id=str(chainPK))
```
The original code handles the case where there is no MicroServiceChain by returning None - this probably should as well.

mistydemeo replied:
Actually, is that okay? Returning `None` has the effect of creating an instance with no attributes, which will blow up only later when something tries to access one of its standard attributes. It's probably better to raise something from within the constructor (even if `MicroServiceChain.DoesNotExist` is not the ideal exception to allow to bubble up)

Hwesta commented on src/MCPServer/lib/linkTaskManagerDirectories.py:
```diff
@@ -28,41 +28,27 @@
+        filterFileEnd = stc.filter_file_end
+        filterFileStart = stc.filter_file_start
+        filterSubDir = stc.filter_subdir
+        self.requiresOutputLock = stc.requires_output_lock
```
This is now a bool, not a string.  Is this a problem?  Same question for other places that use StandardTaskConfig

mistydemeo replied:
Should be fine! This was being fetched as an integer before, and Python treats `if 0/if 1` and `if False/if True` the same.

Hwesta replied:
Ok.  Wasn't sure if it was getting fetched as `"0"` and `"1"`

Hwesta commented on src/MCPServer/lib/linkTaskManagerFiles.py:
```diff
@@ -45,40 +45,39 @@ def __init__(self, jobChainLink, pk, unit):
+        # so they need to be bytestrings here
+        filterFileEnd = str(stc.filter_file_end) if stc.filter_file_end else ''
+        filterFileStart = str(stc.filter_file_start) if stc.filter_file_start else ''
+        filterSubDir = str(stc.filter_subdir) if stc.filter_subdir else ''
```
Should filterFoo get the same treatment in all the other places that query StandardTaskConfig?

mistydemeo replied:
We only use those properties here and in linkTaskManagerDirectories. I don't think ltmDirectories can get unusual paths like we can in ltmFiles.

(The properties are also assigned, but never used, in ltmGetMicroserviceGeneratedListInStdOut)

Hwesta commented on src/MCPServer/lib/unit.py:
```diff
@@ -34,7 +78,30 @@ def setMagicLink(self,link, exitStatus=""):
     def setVariable(self, variable, variableValue, microServiceChainLink):
-        return
+        if not variableValue:
+            variableValue = ""
```
Set microServiceChainLink to None if not set?

mistydemeo replied:
I think it's valid to have an empty variable value, but not valid to have a variable with no microServiceChainLink. I see `null=True` in the model definition, but I'm not sure the original code allowed that?

Hwesta commented on src/MCPServer/lib/unit.py:
```diff
@@ -34,7 +78,30 @@ def setMagicLink(self,link, exitStatus=""):
+        if not variableValue:
+            variableValue = ""
+        vars = UnitVariable.objects.filter(unittype=self.unitType,
+                                           unituuid=self.UUID,
```
Continuing with 'Why has Python taken all the good variable names?', `vars` is a builtin.

mistydemeo replied:
😭

Hwesta commented on src/archivematicaCommon/lib/databaseFunctions.py:
```diff
@@ -20,84 +20,181 @@
+    return str("{:10.10f}".format(float(ret)))
+
+def insertIntoFiles(fileUUID, filePath, enteredSystem=getUTCDate(), transferUUID="", sipUUID="", use="original"):
+    """
```
Change `enteredSystem=getUTCDate` to `enteredSystem=None` and check for None to set enteredSystem.  Function signatures are evaluated when the file is loaded, so getUTCDate() will only be evaluated once when the file is loaded; likely not what we want.

Same for other functions that have getUTCDate() in the function declaration.

mistydemeo replied:
Yeah, I'm not happy this is a common thing. Mind if I make it a followup PR to make this change? I'm focusing on not changing functionality in this one.

mistydemeo replied:
Addressed in https://github.com/artefactual/archivematica/pull/162.

Hwesta commented on src/archivematicaCommon/lib/databaseFunctions.py:
```diff
@@ -20,84 +20,181 @@
-         eventDetail, eventOutcome, eventOutcomeDetailNote, agent)
-    )
+    if not eventIdentifierUUID:
+        eventIdentifierUUID = (uuid.uuid4())
```
```str(uuid.uuid4())```?

Hwesta replied:
Should be `str(uuid.uuid4())`?

mistydemeo replied:
Good catch!

Hwesta commented on src/archivematicaCommon/lib/databaseFunctions.py:
```diff
@@ -20,84 +20,181 @@
+    if not sourceFileUUID:
+        raise ValueError("sourceFileUUID must be specified")
+    if not derivedFileUUID:
+        raise ValueError("derivedFileUUID must be specified")
```
If they're required, then don't make the optional in the function definition?

mistydemeo replied:
We already make heavy use of calling these as kwargs - rather than change the function signature and every place it's called, I thought I'd just verify the arguments instead. I could make that change in a separate commit though if you wanted.

Hwesta replied:
All python functions can be called by specifying their args in order, or by naming them regardless of how the parameter is defined (http://www.diveintopython.net/power_of_introspection/optional_arguments.html) so it shouldn't affect the places this is called.

Probably worthwhile to keep the checks to catch things like `""` since those errors are almost certainly better than what Django generates if the constraint fails.

mistydemeo replied:
I'll punt that to a new PR to avoid increasing the surface area of this one, but let's make those changes.

Hwesta commented on src/archivematicaCommon/lib/databaseFunctions.py:
```diff
@@ -20,84 +20,181 @@
+    derivation = Derivation(source_file_id=sourceFileUUID,
+                            derived_file_id=derivedFileUUID,
+                            event_id=relatedEventUUID)
+    derivation.save()
```
Minor, but `Derivation.objects.create(source_file_id=sourceFileUUID, derived_file_id=derivedFileUUID, event_id=relatedEventUUID)` does the same thing without having to remember to save it.

mistydemeo replied:
Thanks, I thought I'd caught all of these

Hwesta commented on src/archivematicaCommon/lib/databaseFunctions.py:
```diff
@@ -106,53 +203,72 @@ def logTaskCreatedSQL(taskManager, commandReplacementDic, taskUUID, arguments):
+    """
     print "Logging task output to db", task.UUID
     taskUUID = task.UUID.__str__()
     exitCode = task.results["exitCode"].__str__()
```
This is a BigIntegerField - should it be converted to a str?

mistydemeo replied:
I have a feeling it was always being coerced into a string and then back into an integer in the past; could adjust it in a followup PR to not do that?

Hwesta commented on src/archivematicaCommon/lib/databaseFunctions.py:
```diff
@@ -106,53 +203,72 @@ def logTaskCreatedSQL(taskManager, commandReplacementDic, taskUUID, arguments):
-    )
-
-def fileWasRemoved(fileUUID, utcDate=databaseInterface.getUTCDate(), eventDetail = "", eventOutcomeDetailNote = "", eventOutcome=""):
+    job = Job(jobuuid=job.UUID,
```
Redefinition of variable name - job is both the parameter jobChainLink and the newly created Job.

mistydemeo replied:
Ouch, good catch. I'll just not assign the new job to a variable since it's never used/returned.

Hwesta commented on src/archivematicaCommon/lib/databaseFunctions.py:
```diff
@@ -106,53 +203,72 @@ def logTaskCreatedSQL(taskManager, commandReplacementDic, taskUUID, arguments):
+              sipuuid=unitUUID,
+              currentstep="Executing command(s)",
+              unittype=job.unit.__class__.__name__,
+              microservicegroup=str(job.microserviceGroup),
```
Is this already a str?

mistydemeo replied:
Not sure - the old code also always called `str()` on it though.

Hwesta commented on src/archivematicaCommon/lib/elasticSearchFunctions.py:
```diff
@@ -798,8 +797,9 @@ def connect_and_remove_transfer_files(uuid, unit_type=None):
-        sql = "SELECT transferUUID FROM Files WHERE transferUUID=%s OR sipUUID=%s"
-        transfers = {f[0] for f in databaseInterface.queryAllSQL(sql, (uuid, uuid))}
+        condition = Q(transfer_id=uuid) | Q(sip_id=uuid)
+        File.objects.filter(condition).values_list('transfer_id')
```
Can be removed.

Hwesta commented on src/archivematicaCommon/lib/fileOperations.py:
```diff
@@ -155,20 +155,21 @@ def updateFileLocation2(src, dst, unitPath, unitIdentifier, unitIdentifierType,
+    files = File.objects.filter(**kwargs)
+    count = files.count()
+    if count != 1:
+        print >> sys.stderr, 'ERROR: file information not found:', count, "rows for arguments:", repr(kwargs)
```
Use .get() and catch DoesNotExist and MultipleResults instead of counting?  Does mean you can't use .update below, but I think it's clearer.

mistydemeo replied:
Updated.

Hwesta commented on src/archivematicaCommon/tests/test_database_functions.py:
```diff
@@ -0,0 +1,66 @@
+import pytest
+
+
+class TestDatabaseFunctions(TestCase):
```
:D :clap:

mistydemeo commented:
Okay - addressed the comments!

mistydemeo commented:
Rebased on top of qa/1.x.

Hwesta commented on src/dashboard/src/main/models.py:
```diff
@@ -691,3 +837,45 @@ class Meta:
+    id = UUIDPkField()
+    watched_directory_path = models.TextField(null=True, blank=True, db_column='watchedDirectoryPath')
+    chain = models.ForeignKey('MicroServiceChain', null=True, db_column='chain')
+    only_act_on_directories = models.IntegerField(default=1, db_column='onlyActOnDirectories')
```
[Null]BooleanField?

mistydemeo replied:
I'll swap it to `BooleanField`; we've never used `NULL` values in this despite it being allowed in the schema.

Hwesta commented:
Whew!  I think this is ready to go in and be tested in reality. :+1:

Commit:
eraseElasticSearchIndexes: set DJANGO_SETTINGS_MODULE

mistydemeo commented:
Pushed a fix for the eraseElasticSearchIndexes script. Will give this a round of testing, then merge to qa/1.x.
